File: output/GPT3.5-Turbo/xCodeEval_4000/Java/Go/da82c93137395dba5e4c896d5d694179.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

var (
	inf     = int(1e8)
	modular = 998244353
	power   = Power{modular}
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, m, fW, hW int
	fmt.Fscan(reader, &n)
	fmt.Fscan(reader, &m)
	a := make([]int, n)
	count := make([]int, 2)
	for i := 0; i < n; i++ {
		fmt.Fscan(reader, &a[i])
		count[a[i]%2]++
	}
	w := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(reader, &w[i])
		if a[i]%2 == 0 {
			hW += w[i]
		} else {
			fW += w[i]
		}
	}

	inv := make([]int, m*2+1)
	for i := 0; i < m*2+1; i++ {
		inv[i] = power.inverse(hW + fW + i - m)
	}

	f := make([][]int, m+1)
	h := make([][]int, m+1)
	for i := 0; i <= m; i++ {
		f[i] = make([]int, m+1)
		h[i] = make([]int, m+1)
	}

	for i := 0; i <= m; i++ {
		for j := 0; j <= m; j++ {
			f[i][j] = -1
			h[i][j] = -1
		}
	}

	for i := 0; i <= m; i++ {
		f[i][0] = 1
		h[i][0] = 1
	}

	for i := 0; i < n; i++ {
		if a[i]%2 == 1 {
			fmt.Fprintln(writer, mul(w[i], f(0, m)))
		} else {
			fmt.Fprintln(writer, mul(w[i], h(0, m)))
		}
	}
}

func f(i, k int) int {
	if f[i][k] == -1 {
		j := (m - k) - i
		fixI := plus(i, fW)
		fixJ := plus(hW, -j)
		f[i][k] = 0
		inv := inv[i-j+m]
		f[i][k] = plus(f[i][k], mul(mul(plus(fixI, 1), inv), f(i+1, k-1)))
		f[i][k] = plus(f[i][k], mul(mul(fixJ, inv), f(i, k-1)))
	}
	return f[i][k]
}

func h(i, k int) int {
	if h[i][k] == -1 {
		j := (m - k) - i
		fixI := plus(i, fW)
		fixJ := plus(hW, -j)
		h[i][k] = 0
		inv := inv[i-j+m]
		h[i][k] = plus(h[i][k], mul(mul(plus(fixJ, -1), inv), h(i, k-1)))
		h[i][k] = plus(h[i][k], mul(mul(fixI, inv), h(i+1, k-1)))
	}
	return h[i][k]
}

func plus(a, b int) int {
	return (a + b) % modular
}

func mul(a, b int) int {
	return (a * b) % modular
}

type Power struct {
	modular int
}

func (p Power) pow(x int, n int) int {
	if n == 0 {
		return 1
	}
	r := p.pow(x, n>>1)
	r = mul(r, r)
	if n&1 == 1 {
		r = mul(r, x)
	}
	return r
}

func (p Power) inverse(x int) int {
	return p.pow(x, p.modular-2)
}

File: output/GPT3.5-Turbo/xCodeEval_4000/Java/C/da82c93137395dba5e4c896d5d694179.c
#include <stdio.h>
#include <stdlib.h>

#define MAXN 200005
#define MOD 998244353

typedef long long ll;

ll power(ll a, ll b) {
    ll res = 1;
    while (b > 0) {
        if (b & 1) {
            res = (res * a) % MOD;
        }
        a = (a * a) % MOD;
        b >>= 1;
    }
    return res;
}

ll inverse(ll a) {
    return power(a, MOD - 2);
}

ll mul(ll a, ll b) {
    return (a * b) % MOD;
}

ll add(ll a, ll b) {
    return (a + b) % MOD;
}

ll sub(ll a, ll b) {
    return (a - b + MOD) % MOD;
}

ll f[MAXN][2];
ll h[MAXN][2];
ll inv[MAXN * 2 + 1];

void solve() {
    int n, m;
    scanf("%d %d", &n, &m);
    int a[MAXN];
    int count[2] = {0};
    for (int i = 0; i < n; i++) {
        scanf("%d", &a[i]);
        count[a[i] % 2]++;
    }
    int w[MAXN];
    int fW = 0, hW = 0;
    for (int i = 0; i < n; i++) {
        scanf("%d", &w[i]);
        if (a[i] % 2 == 0) {
            hW += w[i];
        } else {
            fW += w[i];
        }
    }

    for (int i = 0; i <= m * 2; i++) {
        inv[i] = inverse(hW + fW + i - m);
    }

    for (int i = 0; i <= m; i++) {
        f[i][0] = 1;
        h[i][0] = 1;
    }

    for (int i = 0; i < n; i++) {
        if (a[i] % 2 == 1) {
            printf("%lld\n", mul(w[i], f[0][m]));
        } else {
            printf("%lld\n", mul(w[i], h[0][m]));
        }
    }
}

int main() {
    solve();
    return 0;
}
